src/java.base/share/classes/java/lang/Thread.java

```diff
@@ -1588,7 +1604,10 @@ void start(ThreadContainer container) {
     public void run() {
         Runnable task = holder.task;
         if (task != null) {
+            Object bindings = scopedValueBindings();
+            ensureMaterializedForStackWalk(bindings);
             task.run();
+            Reference.reachabilityFence(bindings);
```

Comment:
I wonder. The pattern I'm using here is based on `AccessController.executePrivileged`, which doesn't have the `finally` clause. Perhaps I should add one here anyway.

Reply:
I hope it doesn't matter.  There is an example in the reachabilityFence javadocs where it does not use finally.  For it to matter, I think the compiler would need to inline through run() and prove that it can throw an exception, but I don't think that's how the JIT compilers currently implement reachabilityFence.  I suppose a finally shouldn't hurt, however.